src/main/java/part3/exercise/ComposeCachingDataStorage.java

```diff
@@ -21,7 +21,6 @@ public ComposeCachingDataStorage(CachingDataStorage<K1, T1> storage1,
     @Override
     public OutdatableResult<T2> getOutdatable(K1 key) {
 
-        return new OutdatableResult<T2>(storage2.get(storage1.get(key).thenApply(mapping).join()),
-                storage2.getOutdatable(storage1.get(key).thenApply(mapping).join()).getOutdated());
+        return storage2.getOutdatable(storage1.get(key).thenApply(mapping).join());

```

Comment:
What if `storage1.get(key)` expires faster?
Note that you have similar errors in other classes.